packages/driver/src/cypress/command.ts

```diff
@@ -33,9 +33,39 @@ export class $Command {
     return this
   }
 
+  snapshotLogs () {
+    this.get('logs').forEach((log) => {
+      if (!log.get('snapshots')) {
+        log.snapshot()
+      }
+    })
+  }
+
   finishLogs () {
-    // finish each of the logs we have
-    return _.invokeMap(this.get('logs'), 'finish')
+    // TODO: Investigate whether or not we can reuse snapshots between logs
+    // that snapshot at the same time
```

Comment:
Interesting idea. How would this work? My first thought is use a MutationObserver to prevent duplicating snapshots when nothing has changed in the DOM?

Reply:
I was thinking about doing it a little bit less intelligently, and just noting if we're taking multiple snapshots synchronously - in a couple of places we do `command.get('logs').forEach()` sorts of things, where we iterate through commands and / or logs and snapshot all of them.

If we're looping like that without a chance for external JS code to execute, then we know the snapshots can be shared. Mutation observer would be a more intelligent way to do it - just keep a pointer to the most recent snapshot, and invalidate it whenever the DOM mutates.

Dunno, just a thing I was thinking about as I looked at this code. It's come up a couple of times (with Emily looking at memory usage, and also Brian brought up the idea at one point).

Reply:
Also interested in solving the "snapshots are big and memory hungry" problem. Another dumb way (I guess someone thought of this and dismissed it for reasons I hope I'm about to learn) is just to serialize the snapshots and compare them before/after, eg [using XMLSerializer](https://developer.mozilla.org/en-US/docs/Web/API/XMLSerializer).

As an aside, when we talk about memory usage, is the issue primarily **browser** memory usage (this is my understanding). If so, the other solution I've seen floated (also talked about) would just be saving them in some database (sqlite?) on the server and shuttling them back/forth as needed via a websocket. 

The obvious benefit of this is less complexity - just consumes a lot more data. Potential positive would be you can keep the snapshots for use later on (could be useful for many cool things, like replaying something that happened on CI in the Cloud or locally...).

Reply:
I'd prefer to move in the opposite direction, to shuttle *less* over the air between app <-> server, rather than more.

Serializing is expensive, even if we don't retain it in memory. Repeatedly snapshotting and throwing away the snapshots if they're identical is a great way to keep CPU manufacturers in business. :P

Reply:
You'd track the diffs using MutationObserver and only serialize the patches that describe the minimum set of changes, and then rebuild them on demand to a particular state / time